types/jest/index.d.ts

```diff
@@ -1019,7 +1027,7 @@ declare namespace jest {
          * This is particularly useful for ensuring expected objects have the right structure.
          */
         // tslint:disable-next-line: no-unnecessary-generics
-        toStrictEqual<E = any>(expected: E): R;
+        toStrictEqual<E extends T>(expected: E): R;
```

Comment:
We can have something like...
```typescript
declare const expect: <A>(a: A) => ({ toStrictEqual: <E extends ExactWith<A, E>>(e: E) => void })

type ExactWith<A, E> =
  IsPlainObject<A> extends false ? A :
  & { [K in keyof A]: ExactWith<A[K], K extends keyof E ? E[K] : never> }
  & { [K in Exclude<keyof E, keyof A>]: never }

type IsPlainObject<T> =
  T extends object
    ? T extends (...a: any[]) => any
      ? false
      : true
    : false
```
[Playground](https://www.typescriptlang.org/play?ts=true&t=true#code/CYUwxgNghgTiAEYD2A7AzgF3iAHgB3AwC54AeAQQD4AKKE8gSngF5L5qBveDJAZQxgBLMBgCiARwCuUCCVKjsODCBTA08UTigiA6oIwALCgBoNlGiBKimreADckg4PAC+DAFDuMATwIatuvpG5KaibMzu8PAAkmgACtCCKADyAEYAVoQUbLjKquoAZjJoCAD88OTwRJHwAGTwXADaANLwSfAA1iDeSAUVALpWARh6hhQt-aatuSpqnd29GvDlohNV8CggdiAwbC419U2t7ZqQkqCkXT19oqZXi1SDG1s7rp4+frEJUElpmSKkAAq4RqgMUeTmSAyhBqUXKYJm+XYADpUXR4FAUN5Gv0bGxMd5YVFlvAihASkSoiQBJIQESSGSKe5cAQRJwMSQAERQTmuBjInj8IQiCTSCDs9Hc3luZn4QgSrk8vkCvgCYRiKQyBXwKWmVJc1LSjws+VcSVKtwqoXq0Vas0kABMeoNRs8JrZ9p1POdDUQXLA0uVgrVIs14s9uvg+t9YH9gZl7ow2sj0a4sZ1AdcQdVwo1YuT3qjJDTcdMwC5wHjxrlHo5Xs5PpLGarVpDebtdZTxb98AAzK5TJZ4ABGPnuIA)

Reply:
Sorry for misunderstanding, but what problem does this code solve? As I can see from the playground, it checks for additional properies, so `expect({ a: "a" }).toStrictEqual({ a: "a", b: "b" })` will fail. But can't this be achieved with just `toStrictEqual(expected: T): R`?

Also, your example is too restrictive. Consider this example:
```
interface Animal { name: string }
interface Dog extends Animal { breed: string }
const dog: Dog = { name: 'Rex', breed: 'chihuahua' }
const animals: Animal[] = [ dog ]
expect(animals[0]).toStrictEqual({ name: 'Rex', breed: 'chihuahua' })
```

Reply:
Sorry for misunderstanding, but what problem does this code solve? As I can see from the playground, it checks for additional properies, so `expect({ a: "a" }).toStrictEqual({ a: "a", b: "b" })` will fail. But can't this be achieved with just `toStrictEqual(expected: T): R`?

Also, your example is too restrictive. Consider this example:
```
interface Animal { type: string }
interface Dog extends Animal { breed: string }
const dog: Dog = { type: 'dog', breed: 'chihuahua' }
const animals: Animal[] = [ dog ]
expect(animals[0]).toStrictEqual({ type: 'dog', breed: 'chihuahua' })
```

Reply:
(afk) your version won't check for deep equalities and mine would as the playground shows.
Although as you pointed out the actual runtime value can be wider than the type hence we shouldn't have my version. Apart from that I think this would be only work well for literals in other cases I feel it would be a footgun to it's not a great idea anyway. Probably I was a little drunk with type gymnastics when I suggested this xD.
So tbc I'm taking back this suggestion.

Reply:
(on keyboard now) I know this is resolved by realized this part - "your version won't check for deep equalities and mine would" - is not clear in the above playground. So here's a version which would show the difference.

```typescript
declare const expect: <A>(a: A) => ({
  toStrictEqualMine: <E extends ExactWith<A, E>>(e: E) => void,
  toStrictEqualYours: (e: A) => void,
})

type ExactWith<A, E> =
  IsPlainObject<A> extends false ? A :
  & { [K in keyof A]: ExactWith<A[K], K extends keyof E ? E[K] : never> }
  & { [K in Exclude<keyof E, keyof A>]: never }

type IsPlainObject<T> =
  T extends object
    ? T extends (...a: any[]) => any
      ? false
      : true
    : false

let a = { a: "a", b: { c: "c" } }
let b = { a: "a", b: { c: "c", e: 1 } }

expect(a).toStrictEqualMine(b) // error
expect(a).toStrictEqualYours(b) // no error
```

[Playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3iAHgB3AwC54AeAQQD4AKKE8gSngF5L5qBvAKHngyQDKGGAEswGAKIBHAK5QIAWREoQJUhOw4MIFMDTwJOKOIDqIjAAsKAGgOUaqg01bwAbkhHBrPPoOFjJWXkATSQZGDQSakdGFjZ3T28AXwYuLgwATwIDI1NzK3JbCTZmHwBJNAAFaGUAeQAjACtCCjZcbV19ADN5NAQAfnhyeCIfADJ4DngAbQBpeGV4AGsQDKQuoYBdEkNjDDNLCjnN23n2nT1l1fWDeEGJY5H4FVcQGDYk8cmZ+cXDSBkoFIKzWGwkthBNyo22eIFeMHgn3SWQQFWqUDqTRaABUSj5sZoOpckFjxD5eIMCedOuwAHT0ujwKAoDLTTbONjMjLk3h3eA9CB9Hm8EjCGQgHkkAVCrgQEBYKAsb6MgBEUBVtnqJCmYBIKrAKsRiNl8vg9SVU1V6s12sQeoNtkcAEYjUjcARxLQGLT+EJROJpHJFMoQNR6kwAPQR7AwGBIGBcd2EL0+vz+wJB0LhNBhyPRlBIGNxmBAA). The previous version doesn't show the difference probably because they were literals and/or because of contextual inference idk.
Anyway, I just wanted to clarify how my version is different. Still isn't workable.